games/twenty_forty_eight/action.py

```diff
@@ -0,0 +1,79 @@
+from enum import Enum, auto
+from dataclasses import dataclass
+from base.action import Action
+from copy import deepcopy
+from games.twenty_forty_eight.state import TwentyFortyEightState
+import numpy as np
+from numpy.typing import NDArray
+
+
+class Direction(Enum):
+    RIGHT = auto()
+    LEFT = auto()
+    UP = auto()
+    DOWN = auto()
+
+
+@dataclass
+class TwentyFortyEightPlayerAction(Action):
+    direction: Direction
+
+    def apply(self, state: TwentyFortyEightState) -> TwentyFortyEightState:
+        moves = {
+            Direction.LEFT: self._move_left,
+            Direction.RIGHT: self._move_right,
+            Direction.UP: self._move_up,
+            Direction.DOWN: self._move_down
+        }
+        new_board = moves[self.direction](deepcopy(state.board))
+        return TwentyFortyEightState(board=new_board)
+
+    def _move_left(self, board: NDArray) -> NDArray:
+        board = self._compress(board)
+        board = self._merge(board)
+        return self._compress(board)
+
+    def _move_right(self, board: NDArray) -> NDArray:
+        board = self._move_left(np.flip(board, axis=1))
+        return np.flip(board, axis=1)
+
+    def _move_up(self, board: NDArray) -> NDArray:
+        new_board = self._move_left(board.transpose()).transpose()
+        return new_board
+
+    def _move_down(self, board: NDArray) -> NDArray:
+        new_board = self._move_right(board.transpose()).transpose()
+        return new_board
+
+    def _compress(self, board: NDArray) -> NDArray:
```

Comment:
Trochę nie rozumiem tej metody. 

Wygląda tak, jakbyś tworzyła planszę wypełnianą zerami, a potem przenosiłaś do tej planszy wartości z planszy oryginalnej które nie są równe 0.

Czyli coś w stylu

```python
compressed_board = np.zeros(board.shape)
non_zero_mask = board != 0
compressed_board[non_zero_mask] = board[non_zero_mask]
```

Reply:
To jest przesuwanie klocków w lewo. Niezerowe wartości zmieniają pozycję np rząd [0 1 0 4] zamienia się w [1 4 0 0]

Reply:
Fakt. Źle popatrzyłem.